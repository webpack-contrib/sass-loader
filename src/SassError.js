class SassError extends Error {
  constructor(sassError) {
    super();

    this.name = "SassError";
    // TODO remove me in the next major release
    this.originalSassError = sassError;

    if (
      typeof sassError.line !== "undefined" ||
      typeof sassError.column !== "undefined"
    ) {
      this.loc = {
        line: sassError.line,
        column: sassError.column,
      };
    }

    // Keep original error if `sassError.formatted` is unavailable
    this.message = `${this.name}: ${
      typeof this.originalSassError.message !== "undefined"
        ? this.originalSassError.message
        : this.originalSassError
    }`;

    if (this.originalSassError.formatted) {
      this.message = `${this.name}: ${this.originalSassError.formatted.replace(
        /^Error: /,
        ""
      )}`;

      // Instruct webpack to hide the JS stack from the console.
      // Usually you're only interested in the SASS stack in this case.
      this.hideStack = true;

      Error.captureStackTrace(this, this.constructor);
    }
  }
}

export default SassError;
