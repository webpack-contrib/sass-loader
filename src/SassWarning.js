class SassWarning extends Error {
  constructor(warning, options) {
    super(warning);

    this.name = "SassWarning";
    this.hideStack = true;

    if (options.span) {
      this.loc = {
        line: options.span.start.line,
        column: options.span.start.column,
      };
    }
  }
}

export default SassWarning;
