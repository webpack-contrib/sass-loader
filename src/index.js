import url from "url";
import path from "path";

import schema from "./options.json";
import {
  getSassImplementation,
  getSassOptions,
  getWebpackImporter,
  getModernWebpackImporter,
  getCompileFn,
  normalizeSourceMap,
  errorFactory,
} from "./utils";

/**
 * The sass-loader makes node-sass and dart-sass available to webpack modules.
 *
 * @this {object}
 * @param {string} content
 */
async function loader(content) {
  const options = this.getOptions(schema);
  const callback = this.async();

  let implementation;

  try {
    implementation = getSassImplementation(this, options.implementation);
  } catch (error) {
    callback(error);

    return;
  }

  const useSourceMap =
    typeof options.sourceMap === "boolean" ? options.sourceMap : this.sourceMap;
  const sassOptions = await getSassOptions(
    this,
    options,
    content,
    implementation,
    useSourceMap
  );
  const shouldUseWebpackImporter =
    typeof options.webpackImporter === "boolean"
      ? options.webpackImporter
      : true;

  if (shouldUseWebpackImporter) {
    const isModernAPI = options.api === "modern";

    if (!isModernAPI) {
      const { includePaths } = sassOptions;

      sassOptions.importer.push(
        getWebpackImporter(this, implementation, includePaths)
      );
    } else {
      sassOptions.importers.push(
        getModernWebpackImporter(this, implementation)
      );
    }
  }

  let compile;

  try {
    compile = getCompileFn(implementation, options);
  } catch (error) {
    callback(error);
    return;
  }

  let result;

  try {
    result = await compile(sassOptions, options);
  } catch (error) {
    // There are situations when the `file`/`span.url` property do not exist
    // Modern API
    if (error.span && typeof error.span.url !== "undefined") {
      this.addDependency(url.fileURLToPath(error.span.url));
    }
    // Legacy API
    else if (typeof error.file !== "undefined") {
      // `node-sass` returns POSIX paths
      this.addDependency(path.normalize(error.file));
    }

    callback(errorFactory(error));

    return;
  }

  let map =
    // Modern API, then legacy API
    result.sourceMap
      ? result.sourceMap
      : result.map
      ? JSON.parse(result.map)
      : null;

  // Modify source paths only for webpack, otherwise we do nothing
  if (map && useSourceMap) {
    map = normalizeSourceMap(map, this.rootContext);
  }

  // Modern API
  if (typeof result.loadedUrls !== "undefined") {
    result.loadedUrls
      .filter((url) => url.protocol === "file:")
      .forEach((includedFile) => {
        const normalizedIncludedFile = url.fileURLToPath(includedFile);

        // Custom `importer` can return only `contents` so includedFile will be relative
        if (path.isAbsolute(normalizedIncludedFile)) {
          this.addDependency(normalizedIncludedFile);
        }
      });
  }
  // Legacy API
  else if (
    typeof result.stats !== "undefined" &&
    typeof result.stats.includedFiles !== "undefined"
  ) {
    result.stats.includedFiles.forEach((includedFile) => {
      const normalizedIncludedFile = path.normalize(includedFile);

      // Custom `importer` can return only `contents` so includedFile will be relative
      if (path.isAbsolute(normalizedIncludedFile)) {
        this.addDependency(normalizedIncludedFile);
      }
    });
  }

  callback(null, result.css.toString(), map);
}

export default loader;
