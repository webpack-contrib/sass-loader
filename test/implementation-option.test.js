import nodeSass from "node-sass";
import dartSass from "sass";
// eslint-disable-next-line import/no-namespace
import * as sassEmbedded from "sass-embedded";

import { isSupportedFibers } from "../src/utils";

import {
  compile,
  getCodeFromBundle,
  getCompiler,
  getErrors,
  getImplementationByName,
  getImplementationsAndAPI,
  getTestId,
  getWarnings,
} from "./helpers";

jest.setTimeout(30000);

let Fiber;
const implementations = [...getImplementationsAndAPI(), "sass_string"];

describe("implementation option", () => {
  beforeAll(async () => {
    if (isSupportedFibers()) {
      const { default: fibers } = await import("fibers");

      Fiber = fibers;
    }
  });

  beforeEach(() => {
    if (isSupportedFibers()) {
      // The `sass` (`Dart Sass`) package modify the `Function` prototype, but the `jest` lose a prototype
      Object.setPrototypeOf(Fiber, Function.prototype);
    }
  });

  implementations.forEach((item) => {
    let implementationName;
    let implementation;
    let api;

    if (typeof item === "string") {
      implementationName = item;
      implementation = getImplementationByName(implementationName);
      api = "legacy";
    } else {
      ({ name: implementationName, api, implementation } = item);
    }

    it(`'${implementationName}', '${api}' API`, async () => {
      const nodeSassSpy = jest.spyOn(nodeSass, "render");
      const dartSassSpy = jest.spyOn(dartSass, "render");
      const dartSassSpyModernAPI = jest.spyOn(dartSass, "compileStringAsync");
      const sassEmbeddedSpy = jest.spyOn(sassEmbedded, "render");
      const sassEmbeddedSpyModernAPI = jest.spyOn(
        sassEmbedded,
        "compileStringAsync"
      );

      const testId = getTestId("language", "scss");
      const options = { api, implementation };
      const compiler = getCompiler(testId, { loader: { options } });
      const stats = await compile(compiler);
      const { css, sourceMap } = getCodeFromBundle(stats, compiler);

      expect(css).toBeDefined();
      expect(sourceMap).toBeUndefined();

      expect(getWarnings(stats)).toMatchSnapshot("warnings");
      expect(getErrors(stats)).toMatchSnapshot("errors");

      if (implementationName === "node-sass") {
        expect(nodeSassSpy).toHaveBeenCalledTimes(1);
        expect(dartSassSpy).toHaveBeenCalledTimes(0);
        expect(dartSassSpyModernAPI).toHaveBeenCalledTimes(0);
        expect(sassEmbeddedSpy).toHaveBeenCalledTimes(0);
        expect(sassEmbeddedSpyModernAPI).toHaveBeenCalledTimes(0);
      } else if (
        implementationName === "dart-sass" ||
        implementationName === "dart-sass_string"
      ) {
        if (api === "modern") {
          expect(nodeSassSpy).toHaveBeenCalledTimes(0);
          expect(dartSassSpy).toHaveBeenCalledTimes(0);
          expect(dartSassSpyModernAPI).toHaveBeenCalledTimes(1);
          expect(sassEmbeddedSpy).toHaveBeenCalledTimes(0);
          expect(sassEmbeddedSpyModernAPI).toHaveBeenCalledTimes(0);
        } else if (api === "legacy") {
          expect(nodeSassSpy).toHaveBeenCalledTimes(0);
          expect(dartSassSpy).toHaveBeenCalledTimes(1);
          expect(dartSassSpyModernAPI).toHaveBeenCalledTimes(0);
          expect(sassEmbeddedSpy).toHaveBeenCalledTimes(0);
          expect(sassEmbeddedSpyModernAPI).toHaveBeenCalledTimes(0);
        }
      } else if (implementationName === "sass-embedded") {
        if (api === "modern") {
          expect(nodeSassSpy).toHaveBeenCalledTimes(0);
          expect(dartSassSpy).toHaveBeenCalledTimes(0);
          expect(dartSassSpyModernAPI).toHaveBeenCalledTimes(0);
          expect(sassEmbeddedSpy).toHaveBeenCalledTimes(0);
          expect(sassEmbeddedSpyModernAPI).toHaveBeenCalledTimes(1);
        } else if (api === "legacy") {
          expect(nodeSassSpy).toHaveBeenCalledTimes(0);
          expect(dartSassSpy).toHaveBeenCalledTimes(0);
          expect(dartSassSpyModernAPI).toHaveBeenCalledTimes(0);
          expect(sassEmbeddedSpy).toHaveBeenCalledTimes(1);
          expect(sassEmbeddedSpyModernAPI).toHaveBeenCalledTimes(0);
        }
      }

      nodeSassSpy.mockRestore();
      dartSassSpy.mockRestore();
      dartSassSpyModernAPI.mockRestore();
      sassEmbeddedSpy.mockRestore();
      sassEmbeddedSpyModernAPI.mockRestore();
    });
  });

  it("should throw error when unresolved package", async () => {
    const testId = getTestId("language", "scss");
    const options = {
      implementation: "unresolved",
    };
    const compiler = getCompiler(testId, { loader: { options } });
    const stats = await compile(compiler);

    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it("not specify", async () => {
    const nodeSassSpy = jest.spyOn(nodeSass, "render");
    const dartSassSpy = jest.spyOn(dartSass, "render");

    const testId = getTestId("language", "scss");
    const options = {};
    const compiler = getCompiler(testId, { loader: { options } });
    const stats = await compile(compiler);
    const { css, sourceMap } = getCodeFromBundle(stats, compiler);

    expect(css).toBeDefined();
    expect(sourceMap).toBeUndefined();

    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");

    expect(nodeSassSpy).toHaveBeenCalledTimes(0);
    expect(dartSassSpy).toHaveBeenCalledTimes(1);

    nodeSassSpy.mockRestore();
    dartSassSpy.mockRestore();
  });

  it("not specify with legacy API", async () => {
    const nodeSassSpy = jest.spyOn(nodeSass, "render");
    const dartSassSpy = jest.spyOn(dartSass, "render");

    const testId = getTestId("language", "scss");
    const options = {
      api: "legacy",
    };
    const compiler = getCompiler(testId, { loader: { options } });
    const stats = await compile(compiler);
    const { css, sourceMap } = getCodeFromBundle(stats, compiler);

    expect(css).toBeDefined();
    expect(sourceMap).toBeUndefined();

    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");

    expect(nodeSassSpy).toHaveBeenCalledTimes(0);
    expect(dartSassSpy).toHaveBeenCalledTimes(1);

    nodeSassSpy.mockRestore();
    dartSassSpy.mockRestore();
  });

  it("not specify with modern API", async () => {
    const nodeSassSpy = jest.spyOn(nodeSass, "render");
    const dartSassSpy = jest.spyOn(dartSass, "compileStringAsync");

    const testId = getTestId("language", "scss");
    const options = {
      api: "modern",
    };
    const compiler = getCompiler(testId, { loader: { options } });
    const stats = await compile(compiler);
    const { css, sourceMap } = getCodeFromBundle(stats, compiler);

    expect(css).toBeDefined();
    expect(sourceMap).toBeUndefined();

    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");

    expect(nodeSassSpy).toHaveBeenCalledTimes(0);
    expect(dartSassSpy).toHaveBeenCalledTimes(1);

    nodeSassSpy.mockRestore();
    dartSassSpy.mockRestore();
  });

  it("should throw an error on an unknown sass implementation", async () => {
    const testId = getTestId("language", "scss");
    const options = {
      implementation: Object.assign({}, dartSass, {
        info: "strange-sass\t1.0.0",
      }),
    };

    const compiler = getCompiler(testId, { loader: { options } });
    const stats = await compile(compiler);

    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should throw an error when the "info" is unparseable', async () => {
    const testId = getTestId("language", "scss");
    const options = {
      implementation: Object.assign({}, dartSass, { info: "asdfj" }),
    };

    const compiler = getCompiler(testId, { loader: { options } });
    const stats = await compile(compiler);

    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should throw error when the "info" does not exist', async () => {
    const testId = getTestId("language", "scss");
    const options = {
      // eslint-disable-next-line no-undefined
      implementation: Object.assign({}, dartSass, { info: undefined }),
    };

    const compiler = getCompiler(testId, { loader: { options } });
    const stats = await compile(compiler);

    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it("should try to load using valid order", async () => {
    jest.doMock("sass", () => {
      const error = new Error("Some error sass");

      error.code = "MODULE_NOT_FOUND";
      error.stack = null;

      throw error;
    });

    jest.doMock("node-sass", () => {
      const error = new Error("Some error node-sass");

      error.code = "MODULE_NOT_FOUND";
      error.stack = null;

      throw error;
    });

    jest.doMock("sass-embedded", () => {
      const error = new Error("Some error sass-embedded");

      error.code = "MODULE_NOT_FOUND";
      error.stack = null;

      throw error;
    });

    const testId = getTestId("language", "scss");
    const options = {};

    const compiler = getCompiler(testId, { loader: { options } });
    const stats = await compile(compiler);

    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it("should not swallow an error when trying to load a sass implementation", async () => {
    jest.doMock("node-sass", () => {
      const error = new Error("Some error");

      error.code = "MODULE_NOT_FOUND";
      error.stack = null;

      throw error;
    });

    jest.doMock("sass", () => {
      const error = new Error("Some error");

      error.code = "MODULE_NOT_FOUND";
      error.stack = null;

      throw error;
    });

    const testId = getTestId("language", "scss");
    const options = {};

    const compiler = getCompiler(testId, { loader: { options } });
    const stats = await compile(compiler);

    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });
});
