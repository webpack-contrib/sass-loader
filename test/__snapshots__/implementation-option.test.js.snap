// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`implementation option 'dart-sass', 'legacy' API: errors 1`] = `[]`;

exports[`implementation option 'dart-sass', 'legacy' API: warnings 1`] = `[]`;

exports[`implementation option 'dart-sass', 'modern' API: errors 1`] = `[]`;

exports[`implementation option 'dart-sass', 'modern' API: warnings 1`] = `[]`;

exports[`implementation option 'node-sass', 'legacy' API: errors 1`] = `[]`;

exports[`implementation option 'node-sass', 'legacy' API: warnings 1`] = `[]`;

exports[`implementation option 'sass_string', 'legacy' API: errors 1`] = `[]`;

exports[`implementation option 'sass_string', 'legacy' API: warnings 1`] = `[]`;

exports[`implementation option 'sass-embedded', 'legacy' API: errors 1`] = `[]`;

exports[`implementation option 'sass-embedded', 'legacy' API: warnings 1`] = `[]`;

exports[`implementation option 'sass-embedded', 'modern' API: errors 1`] = `[]`;

exports[`implementation option 'sass-embedded', 'modern' API: warnings 1`] = `[]`;

exports[`implementation option not specify with legacy API: errors 1`] = `[]`;

exports[`implementation option not specify with legacy API: warnings 1`] = `[]`;

exports[`implementation option not specify with modern API: errors 1`] = `[]`;

exports[`implementation option not specify with modern API: warnings 1`] = `[]`;

exports[`implementation option not specify: errors 1`] = `[]`;

exports[`implementation option not specify: warnings 1`] = `[]`;

exports[`implementation option should not swallow an error when trying to load a sass implementation: errors 1`] = `
[
  "ModuleBuildError: Module build failed (from ../src/cjs.js):
Some error",
]
`;

exports[`implementation option should not swallow an error when trying to load a sass implementation: warnings 1`] = `[]`;

exports[`implementation option should throw an error on an unknown sass implementation: errors 1`] = `
[
  "ModuleBuildError: Module build failed (from ../src/cjs.js):
Error: Unknown Sass implementation "strange-sass".",
]
`;

exports[`implementation option should throw an error on an unknown sass implementation: warnings 1`] = `[]`;

exports[`implementation option should throw an error when the "info" is unparseable: errors 1`] = `
[
  "ModuleBuildError: Module build failed (from ../src/cjs.js):
Error: Unknown Sass implementation "asdfj".",
]
`;

exports[`implementation option should throw an error when the "info" is unparseable: warnings 1`] = `[]`;

exports[`implementation option should throw error when the "info" does not exist: errors 1`] = `
[
  "ModuleBuildError: Module build failed (from ../src/cjs.js):
Error: Unknown Sass implementation.",
]
`;

exports[`implementation option should throw error when the "info" does not exist: warnings 1`] = `[]`;

exports[`implementation option should throw error when unresolved package: errors 1`] = `
[
  "ModuleBuildError: Module build failed (from ../src/cjs.js):
Error: Cannot find module 'unresolved' from 'src/utils.js'",
]
`;

exports[`implementation option should throw error when unresolved package: warnings 1`] = `[]`;

exports[`implementation option should try to load using valid order: errors 1`] = `
[
  "ModuleBuildError: Module build failed (from ../src/cjs.js):
Some error sass",
]
`;

exports[`implementation option should try to load using valid order: warnings 1`] = `[]`;
